client/rpc-spec-v2/src/chain_head/chain_head.rs

```diff
@@ -0,0 +1,477 @@
+		+ CallApiAt<Block>
+		+ StorageProvider<Block, BE>
+		+ 'static,
+{
+	fn chain_head_unstable_follow(
+		&self,
+		mut sink: SubscriptionSink,
+		runtime_updates: bool,
+	) -> SubscriptionResult {
+		let sub_id = self.accept_subscription(&mut sink)?;
+		// Keep track of the subscription.
+		self.subscriptions.insert_subscription(sub_id.clone());
+
+		let client = self.client.clone();
+		let subscriptions = self.subscriptions.clone();
+
+		let sub_id_import = sub_id.clone();
+		let stream_import = self
+			.client
+			.import_notification_stream()
+			.map(move |notification| {
+				let new_runtime = generate_runtime_event(
+					&client,
+					runtime_updates,
+					&BlockId::Hash(notification.hash),
+					Some(&BlockId::Hash(*notification.header.parent_hash())),
+				);
+
+				let _ = subscriptions.pin_block(&sub_id_import, notification.hash);
+
+				// Note: `Block::Hash` will serialize to hexadecimal encoded string.
+				let new_block = FollowEvent::NewBlock(NewBlock {
+					block_hash: notification.hash,
+					parent_block_hash: *notification.header.parent_hash(),
+					new_runtime,
+					runtime_updates,
+				});
+
+				if !notification.is_new_best {
+					return stream::iter(vec![new_block])
+				}
+
+				// If this is the new best block, then we need to generate two events.
+				let best_block = FollowEvent::BestBlockChanged(BestBlockChanged {
+					best_block_hash: notification.hash,
+				});
+				stream::iter(vec![new_block, best_block])
+			})
+			.flatten();
+
+		let subscriptions = self.subscriptions.clone();
+		let sub_id_finalized = sub_id.clone();
+
+		let stream_finalized =
+			self.client.finality_notification_stream().map(move |notification| {
+				// We might not receive all new blocks reports, also pin the block here.
+				let _ = subscriptions.pin_block(&sub_id_finalized, notification.hash);
+
+				// The tree route contains the exclusive path from the latest finalized block
+				// to the block reported by the notification. Ensure the finalized block is
+				// properly reported to that path.
+				let mut finalized_block_hashes =
+					notification.tree_route.iter().cloned().collect::<Vec<_>>();
+				finalized_block_hashes.push(notification.hash);
+
+				FollowEvent::Finalized(Finalized {
+					finalized_block_hashes,
+					pruned_block_hashes: notification.stale_heads.iter().cloned().collect(),
+				})
+			});
+
+		let merged = tokio_stream::StreamExt::merge(stream_import, stream_finalized);
+
+		// The initialized event is the first one sent.
+		let finalized_block_hash = self.client.info().finalized_hash;
+		let finalized_block_runtime = generate_runtime_event(
+			&self.client,
+			runtime_updates,
+			&BlockId::Hash(finalized_block_hash),
+			None,
+		);
+
+		let _ = self.subscriptions.pin_block(&sub_id, finalized_block_hash);
+
+		let stream = stream::once(async move {
+			FollowEvent::Initialized(Initialized {
+				finalized_block_hash,
+				finalized_block_runtime,
+				runtime_updates,
+			})
+		})
```

Comment:
> Afterwards, generates one newBlock notification (see below) for each non-finalized block currently in the node's memory (including all forks), then a bestBlockChanged notification. The notifications must be sent in an ordered way such that the parent of each block either can be found in an earlier notification or is the current finalized block.



The spec says that after the `Initialized` event there should be a `NewBlock` event emitted for every block in node memory. I think this is currently missing.

Reply:
Nice catch! I've added the functionality of reporting the "in-memory" blocks by recursively walking the children of the latest finalized block. 
While at it, I also added new testing:
- checked the reported blocks with forks on initialize
- checked that pruned and finalized blocks are reported correctly for the `finalized` event